flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierAligner.java

```diff
@@ -100,21 +119,22 @@ public boolean isBlocked(int channelIndex) {
 	}
 
 	@Override
-	public boolean processBarrier(CheckpointBarrier receivedBarrier, int channelIndex, long bufferedBytes) throws Exception {
+	public void processBarrier(CheckpointBarrier receivedBarrier, int channelIndex) throws Exception {
 		final long barrierId = receivedBarrier.getId();
+		List<Integer> channelsToUnblock = new ArrayList<>();
```

Comment:
In mostly normal cases, this list should not be created beforehand, so it is better to create based on demand. 
I know it is for unification consideration. Another option is that we can migrate partial resume notification inside `releaseBlocksAndResetBarriers` which is involved in batch of channels to be notified.

Then the other cases are all for the single channel notification which can be done inside `#processBarrier`. Then we can define boolean `shouldNotify` beforehand, and to remove unnecessary two `else` conditions below to make the logics seem shorter. And at the end to notify the `channelIndex` if `shouldNotify` set as `false` intermediate steps.

Another potential benefit is to make notification happen before `notifyCheckpoint`. `notifyCheckpoint` would be involve in IO operations in sync process, so it is better to notify light-weight unblock actions beforehand to make data ready earlier.